### PR TITLE
add objective function to define feature visualization process

### PR DESCRIPTION
**Pull request by ResearchDaniel**
**Describe your improvements**
- Added layer-level deep dream visualization and refactored to promote code reusability. Different objectives can now be used to do filter/layer level visualization. 
- Separated image transformation so that it is only performed for loss computation, not for optimization. The standard transformations did not work as expected otherwise.

Using standard_transformation, before/after this PR (`images.initialize_image_ref(299, 299, fft=False, decorrelate=False, seed=1)`)
inceptionv3_optimize_transformed
inceptionv3_optimize_param

### CODE REVIEW

File: luna/featurevis/objectives.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import tensorflow as tf
+
+class FilterObjective(object):
```
Comment (Sparkier) [inline luna/featurevis/objectives.py:3]: I think we should be consistent between channel/filter.

File: docs/source/usage.rst [hunk 2]
```diff
@@ -31,7 +31,9 @@ The following provides an example of how Luna can be used for feature visualizat
         return img
 
     opt_param = featurevis.OptimizationParameters(iterations, learning_rate, optimizer=optimizer)
-    activation, image= featurevis.visualize(image, model, layer="mixed5", optimization_parameters=opt_param, filter_index=30, transformation=my_trans)
+    # Visualize channel 30 within the mixed5 layer
+    objective = featurevis.objectives.FilterObjective(model, layer="mixed5", filter_index=30)
```
Comment (Sparkier) [inline docs/source/usage.rst:35]: channel vs. filter

Should be consistent.

File: luna/featurevis/featurevis.py [hunk 3]
```diff
@@ -22,28 +22,23 @@ class OptimizationParameters():
...
     optimizer: Optional[object]
 
-
 def visualize(
     image,
-    model,
-    layer,
+    objective,
     optimization_parameters,
     filter_index=None,
```
Comment (Sparkier) [inline luna/featurevis/featurevis.py:29]: Can't we remove this now?

File: luna/featurevis/featurevis.py [hunk 4]
```diff
@@ -86,47 +80,78 @@ def visualize(
...
+        tuple: activation and result image for the process.
+    """
+    objective = objectives.LayerObjective(model, layer, regularization)
+    return visualize(image, objective, optimization_parameters, transformation, threshold, minimize)
 
-def gradient_ascent_step(img, model, regularization, optimization_parameters,
-                         minimize, filter_index=None):
-    """Performing one step of gradient ascend.
+def gradient_ascent_step(img, objective, optimization_parameters,
+                         minimize):
```
Comment (Sparkier) [inline luna/featurevis/featurevis.py:149]: Nit: can be one line now.

File: luna/featurevis/images.py [hunk 5]
```diff
@@ -124,11 +126,13 @@ def initialize_image_ref(
...
-    if channels:
-        output = tf.nn.sigmoid(image_f)
-    else:
+    if channels == 3:
         output = to_valid_rgb(
             image_f[..., :3], decorrelate=decorrelate, sigmoid=True)
+    else:
+        output = tf.nn.sigmoid(image_f)
+
+
```
Comment (Sparkier) [inline luna/featurevis/images.py:135]: Blank lines should be removed.

**Review by Sparkier — COMMENTED** [review]
Only minor changes from my side. I think this is the preferred implementation, mine was more of a hot fix.

I will let Fuzhan add her review, however. I think by now she knows the code best.

File: luna/featurevis/featurevis.py (hunk 3 above)
Comment (ResearchDaniel) [reply luna/featurevis/featurevis.py:29]: Good catch

Comment (ResearchDaniel) [reply luna/featurevis/featurevis.py:29]: Ah, it was a merge conflict fix error

File: luna/featurevis/objectives.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        """ Deepdream visualization of a layer, see Mordvintsev et al. 2015.
+            Computes (mean activation)^2 of the input.
+            Args: 
+                model (object): the model to be used for the feature visualization.
+                layer (string): the name of the layer to be used in the visualization.
+                regularization (function): customized regularizers to be applied. Defaults to None.
+        """
+        self.model = get_feature_extractor(model, layer)
+        self.regularization = regularization
+
```
Comment (fuzhanrahmanian) [inline luna/featurevis/objectives.py:57]: Since pylint does not like classes with just one public method, it could be useful to implement a standard __repr__ function:

```suggestion
def __repr__(self) -> str:
        return f"LayerObjective({self.model}, {self.regularization})"
```

Alternatively, we could just deactivate the pylint error at the beginning of the class with:

" # pylint: disable=R0903 "

File: luna/featurevis/objectives.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    """Builds a model that that returns the activation of the specified layer.
+
+    Args:
+        model (object): the model used as a basis for the feature extractor.
+        layer (string): the layer at which to cap the original model.
+    """
+    layer = model.get_layer(name=layer_name)
+    return tf.keras.Model(inputs=model.inputs, outputs=layer.output)
+
+
```
Comment (fuzhanrahmanian) [inline luna/featurevis/objectives.py:86]: Extra white line. I think pylint also does not like this

File: luna/featurevis/objectives.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            Args: 
+                model (object): the model to be used for the feature visualization.
+                layer (string): the name of the layer to be used in the visualization.
+                filter_index (number): the index of the filter to be visualized.
+                regularization (function): customized regularizers to be applied. Defaults to None.
+        """
+        self.model = get_feature_extractor(model, layer)
+        self.filter_index = filter_index
+        self.regularization = regularization
+
```
Comment (fuzhanrahmanian) [inline luna/featurevis/objectives.py:20]: Since pylint does not like classes with just one public method, it could be useful to implement a standard __repr__ function:

```suggestion
def __repr__(self) -> str:
        return f"FilterObjective({self.model},\
                   {self.filter_index}, {self.regularization})"
```

Alternatively, we could just deactivate the pylint error at the beginning of the class with:

" # pylint: disable=R0903 "

**Review by fuzhanrahmanian — CHANGES_REQUESTED** [review]
This is a very good way for distinguishing between Layer and Filter. Like the idea a lot to put it in separate classes. 
I run my local test and the functionality is intact. 

Please review some minor suggestion to make pylint and travis happy.

**Comment by ResearchDaniel** [comment]
@fuzhanrahmanian Thanks for the very helpful comments. I installed pylint and fixed all your suggestions.

File: luna/featurevis/images.py [hunk 9]
```diff
@@ -146,16 +148,6 @@ def fft_image(img, std=None, decay_power=1):
     Returns:
         New image in spatial domain.
     """
-    if tf.compat.v1.keras.backend.image_data_format() == "channels_first":
```
Comment (fuzhanrahmanian) [inline luna/featurevis/images.py:147]: I think that this check is still necessary.
Since the function is accepting a shape that can be either channel first or not (see line 116 or 118), batch, channels, height, width as a general shape becomes incorrect. 

A "channel-last" model, would yield batch, channels, height, width = 1  299 3 3 which is not correct. 
A test with 
`images.initialize_image_ref(299, 299, fft=True, decorrelate=True, seed=1)`
would produce an error.

**Review by fuzhanrahmanian — CHANGES_REQUESTED** [review]
Thank you, Daniel, for the changes. Nice that pylint helps you.
However, my test was not successful, especially regarding the shapes when calling FFT. I think that we should still separate between channel first and channel last when defining the shape.

File: luna/featurevis/images.py (hunk 9 above)
Comment (ResearchDaniel) [reply luna/featurevis/images.py:147]: I had missed to push my the last commit that I think fixes this. Can you verify with the latest changes?

**Review by fuzhanrahmanian — CHANGES_REQUESTED** [review]
With this change, the color seems to disappear almost completely from the result. In mini-lucid, the gradient is calculated with activation and not loss (if you are trying to resemble mini-lucid that is).

I would suggest fixing the gradient override instead of completely removing it. 

Small reminder @ResearchDaniel: the previous issues about the shape of the last review are still open.

File: luna/featurevis/images.py [hunk 10]
```diff
@@ -118,15 +115,13 @@ def initialize_image_ref(
...
+        image_f = fft_image(width, height, channels, std=std)
     else:
         std = std or 0.01
         if seed is not None:
             np.random.seed(seed)
-        image_f = np.random.normal(
-            size=[shape[0], shape[1],
-                  shape[2], shape[3]], scale=std
-        ).astype(np.float32)
+        shape = _to_img_shape(width, height, channels)
```
[the remark this answers is not in the document]
Comment (ResearchDaniel) [reply luna/featurevis/images.py]: Fixed

**Comment by ResearchDaniel** [comment]
> In mini-lucid, the gradient is calculated with activation and not loss (if you are trying to resemble mini-lucid that is)

I don't think that I changed the gradient calculation. I just renamed it

File: luna/featurevis/featurevis.py [hunk 11]
```diff
@@ -153,27 +188,19 @@ def gradient_ascent_step(img, model, regularization, optimization_parameters,
             learning_rate = optimization_parameters.learning_rate or 0.7
             img = img + learning_rate * grads
         else:
-            grads_relu_0 = rg.redirected_relu_grad(img, grads*-1)
-            grads_modified = rg.redirected_relu6_grad(img, grads_relu_0)
-            optimization_parameters.optimizer.apply_gradients(zip([grads_modified], [img]))
+            # Relu gradient overrides do not seem to work,
```
Comment (fuzhanrahmanian) [inline luna/featurevis/featurevis.py:191]: I would suggest keeping the current version of the gradient override as by removing it, the problem still persists (at least with this gradient we are getting some minimal coloration)

**Review by fuzhanrahmanian — CHANGES_REQUESTED** [review]

File: luna/featurevis/featurevis.py (hunk 11 above)
Comment (ResearchDaniel) [reply luna/featurevis/featurevis.py:191]: These are the results I get with/without the current gradient overrides.
**With gradient overrides. Transformations: scale + blur**
test_scale_blur_with_gradient_overrides
**Without gradient overrides. Transformations: scale + blur**
test_scale_blur_without_grad_override
**With gradient overrides. Transformations: default**
test_default_trans_with_gradient_overrides
**Without gradient overrides. Transformations: default**
test_default_transformations_without_gradient_overrides


I think that separating the transformations out of the the optimized input is the reason why the gradient overrides now impact the result (more?).

File: luna/featurevis/objectives.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+            if not callable(self.regularization):
+                raise ValueError("The regularizations need to be a function.")
+            activation_score = self.regularization(
+                activation, activation_score)
+        return -activation_score
+
+    def __repr__(self) -> str:
+        return f"LayerObjective({self.model}, {self.regularization})"
+
+class LayerActivationObjective:
```
Comment (ResearchDaniel) [inline luna/featurevis/objectives.py:90]: @Sparkier

File: luna/featurevis/featurevis.py (hunk 11 above)
Comment (Sparkier) [reply luna/featurevis/featurevis.py:191]: So what is our conclusion? We remove them for now and add them when we have an idea how to make them work?

File: luna/featurevis/objectives.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+    def loss(self, input_image):
+        """Computes the layer loss for the feature visualization process.
+
+        Args:
+            input_image (array): the image that is used to compute the loss.
+            model (object): the model on which to compute the loss.
+        Returns:
+            number: the loss for the specified setting
+        """
+
```
Comment (Sparkier) [inline luna/featurevis/objectives.py:117]: Nit: remove blank line.

Comment (Sparkier) [inline luna/featurevis/objectives.py:117]: Nit: remove blank line.

**Review by Sparkier — APPROVED** [review]
Only thing remaining is the gradient override. I think we should probably remove it for now and make an issue that we address soon to look into it again.

**Review by fuzhanrahmanian — APPROVED** [review]

File: luna/featurevis/featurevis.py (hunk 11 above)
Comment (fuzhanrahmanian) [reply luna/featurevis/featurevis.py:191]: Running the test file, at the moment, the result without gradient override looks like this:
test
 
In my opinion, a "broken" gradient override is still a better result regarding both color and sharpness than something completely blurry but if you both would like to remove it, I am not against it.
